09 - Multithreading/09 - Multithreading.ipynb

```diff
@@ -0,0 +1,1420 @@
+    "print(\"I want to wait for it to end\")\n",
+    "t1.join()  # Comment this, to see the change\n",
+    "print(\"The thread has finished\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Методът `join` спира изпълненето на кода, докато нишката не приключи работата си или не срещне `Exception`. Възможно е да подадем `timeout` параметър на `join` - той индикира колко време да изчакаме нишката да приключи. Важно е да отбележим, че след изтичането на `timeout` параметъра, нишката може да продължи да работи. С методът `is_alive` можем да проверим дали дадена нишка все още работи."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 34,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      ">Going to sleep\n",
+      "Is the thread working ? True\n",
+      "Let's give it some time\n",
+      "Time to get up !\n",
+      "Is the thread working ? True\n",
+      ">Did a really good nap\n",
+      "What about now ? False\n"
+     ]
+    }
+   ],
+   "source": [
+    "from threading import Thread\n",
+    "from time import sleep \n",
+    "\n",
+    "def sleepy_thread():\n",
+    "    print(\">Going to sleep\")\n",
+    "    sleep(20)\n",
+    "    print(\">Did a really good nap\")\n",
+    "\n",
+    "t1 = Thread(target=sleepy_thread)\n",
+    "t1.start()\n",
+    "\n",
+    "print(f\"Is the thread working ? {t1.is_alive()}\")\n",
+    "\n",
+    "print(\"Let's give it some time\")\n",
+    "sleep(2)\n",
+    "\n",
+    "print(\"Time to get up !\")\n",
+    "t1.join(timeout=3)\n",
+    "\n",
+    "print(f\"Is the thread working ? {t1.is_alive()}\")\n",
+    "t1.join()\n",
+    "\n",
+    "print(f\"What about now ? {t1.is_alive()}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Друга важна особеност на `Thread`, е че ако нашата функция връща някакъв резутлат, той няма да ни бъде върнат, ако функцията се изпълнява през thread."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 35,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "res=None, type(res)=<class 'NoneType'>\n",
+      "res=None, type(res)=<class 'NoneType'>\n"
+     ]
+    }
+   ],
+   "source": [
+    "from threading import Thread\n",
+    "\n",
+    "def return_a_value():\n",
+    "    return 5\n",
+    "\n",
+    "t1 = Thread(target=return_a_value)\n",
+    "\n",
+    "res = t1.start()\n",
+    "print(f\"{res=}, {type(res)=}\")\n",
+    "\n",
+    "res = t1.join()\n",
+    "print(f\"{res=}, {type(res)=}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Нека разгледаме следния практически пример - имаме следната функция, която изчислява (до някаква степен) безкрайна сума. Като пример, ще вземем следната сума  $$ \\sum_{k=0}^{\\infty} \\frac{99^k}{(99(k+1))^{97}}  $$\n",
```

Comment:
What sorcery is this
![molq](https://glasnews.bg/news/2020/04/02/generalat-veche-shte-ni-gleda-i-357.jpg)